### PR TITLE
Create GT911 Touchscreen component

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix?

This creates the GT911 touchscreen component as found in the m5paper.

Also used in the Espressif ESP32-S3-Box-3.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3400

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
touchscreen:
  - platform: gt911
    id: gt911_touchscreen
    interrupt_pin:
      number: GPIO3
      ignore_strapping_warning: true

binary_sensor:
  - platform: gt911
    name: Red Button
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nielsnl68** [comment]
ping

**Comment by nielsnl68** [comment]
ping

**Comment by jesserockz** [comment]
> ping

pong

File: esphome/components/gt911/gt911_touchscreen.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+#define ERROR_CHECK(err) \
+  if ((err) != i2c::ERROR_OK) { \
+    ESP_LOGE(TAG, "Failed to communicate!"); \
+    this->status_set_warning(); \
+    return; \
+  }
+
+void IRAM_ATTR HOT Store::gpio_intr(Store *store) {
+  InterruptLock lock;
```
Comment (jesserockz) [inline esphome/components/gt911/gt911_touchscreen.cpp]:
```suggestion
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  this->interrupt_pin_->pin_mode(gpio::FLAG_INPUT | gpio::FLAG_PULLUP);
+  this->interrupt_pin_->setup();
+
+  if (this->write(nullptr, 0) != i2c::ERROR_OK) {
+    ESP_LOGE(TAG, "Failed to communicate!");
+    this->mark_failed();
+    return;
+  }
+
+  this->interrupt_pin_->attach_interrupt(Store::gpio_intr, &this->store_, gpio::INTERRUPT_RISING_EDGE);
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.cpp]:
```suggestion
  this->attach_interrupt_(this->interrupt_pin_,  gpio::INTERRUPT_RISING_EDGE);
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+void GT911Touchscreen::loop() {
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:52]:
```suggestion
void GT911Touchscreen::update_touches() {
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  i2c::ErrorCode err;
+  touchscreen::TouchPoint tp;
+  uint8_t touch_state = 0;
+  uint8_t data[MAX_TOUCHES + 1][8];  // 8 bytes each for each point, plus extra space for the key byte
+
+  if (!this->store_.available)
+    return;
+  this->store_.available = false;
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:60]:
```suggestion
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  err = this->read(&touch_state, 1);
+  ERROR_CHECK(err);
+  this->write(CLEAR_TOUCH_STATE, sizeof(CLEAR_TOUCH_STATE));
+
+  if ((touch_state & 0x80) == 0)
+    return;
+  uint8_t num_of_touches = touch_state & 0x07;
+  bool has_key = (touch_state & 0x10) != 0;
+  if (num_of_touches == 0)
+    this->send_release_();
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:72]:
```suggestion
    return;
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+      case touchscreen::ROTATE_180_DEGREES:
+        tp.x = this->display_width_ - y;
+        tp.y = this->display_height_ - x;
+        break;
+      case touchscreen::ROTATE_270_DEGREES:
+        tp.x = this->display_height_ - y;
+        tp.y = x;
+        break;
+    }
+    this->defer([this, tp]() { this->send_touch_(tp); });
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.cpp]:
```suggestion
    set_raw_touch_posistion_(id, x y);
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    this->mark_failed();
+    return;
+  }
+
+  this->interrupt_pin_->attach_interrupt(Store::gpio_intr, &this->store_, gpio::INTERRUPT_RISING_EDGE);
+}
+
+void GT911Touchscreen::loop() {
+  i2c::ErrorCode err;
+  touchscreen::TouchPoint tp;
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:54]:
```suggestion
  uint8_t id;
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+#define ERROR_CHECK(err) \
+  if ((err) != i2c::ERROR_OK) { \
+    ESP_LOGE(TAG, "Failed to communicate!"); \
+    this->status_set_warning(); \
+    return; \
+  }
+
+void IRAM_ATTR HOT Store::gpio_intr(Store *store) { store->available = true; }
+
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.cpp:25]:
```suggestion
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.h [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+#include "esphome/core/hal.h"
+
+namespace esphome {
+namespace gt911 {
+
+struct Store {
+  volatile bool available;
+
+  static void gpio_intr(Store *store);
+};
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.h:15]:
```suggestion
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.h [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+class GT911ButtonListener {
+ public:
+  virtual void update_button(uint8_t index, bool state) = 0;
+};
+
+class GT911Touchscreen : public touchscreen::Touchscreen, public Component, public i2c::I2CDevice {
+ public:
+  void setup() override;
+  void loop() override;
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.h:25]:
```suggestion
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  void dump_config() override;
+
+  void set_interrupt_pin(InternalGPIOPin *pin) { this->interrupt_pin_ = pin; }
+  void register_button_listener(GT911ButtonListener *listener) { this->button_listeners_.push_back(listener); }
+
+ protected:
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.h:32]:
```suggestion
 protected:
   void update_touches() override;
```

File: esphome/components/gt911/touchscreen/gt911_touchscreen.h [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  InternalGPIOPin *interrupt_pin_;
+  Store store_;
```
Comment (nielsnl68) [inline esphome/components/gt911/touchscreen/gt911_touchscreen.h:34]:
```suggestion
```

File: esphome/components/touchscreen/touchscreen.h [hunk 13]
```diff
@@ -41,6 +41,7 @@ class Touchscreen {
  protected:
   /// Call this function to send touch points to the `on_touch` listener and the binary_sensors.
   void send_touch_(TouchPoint tp);
+  void send_release_();
```
Comment (nielsnl68) [inline esphome/components/touchscreen/touchscreen.h:44]:
```suggestion
```
The new touchscreen handler has this option already.

File: esphome/components/touchscreen/touchscreen.cpp [hunk 14]
```diff
@@ -18,6 +18,11 @@ void Touchscreen::set_display(display::Display *display) {
   }
 }
 
+void Touchscreen::send_release_() {
+  for (auto *listener : this->touch_listeners_)
+    listener->release();
+}
+
```
Comment (nielsnl68) [inline esphome/components/touchscreen/touchscreen.cpp:25]:
```suggestion
```

**Review by nielsnl68 — COMMENTED** [review]
The following changes are based on the new `touchscreen` component. As you can see a lot of work has been taken out of your hands already by the Touchscreen class.

At the moment you do not need to update any of the changes. Until my PR #4596 is merged.

**Pull request merged**